server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,10 +233,158 @@ private void performOnReplicas(
+         * @param primaryRouting primary ShardRouting
+         * @return the determined replication mode.
+         */
+        abstract ReplicationMode determineReplicationMode(
+            final ReplicationModeAwareShardRouting shardRouting,
+            final ShardRouting primaryRouting
+        );
+    }
+
+    /**
+     * This implementation of {@link ReplicationProxy} fans out the replication request to current shard routing if
+     * it is not the primary and has replication mode as {@link ReplicationMode#LOGICAL_REPLICATION}.
+     *
+     * @opensearch.internal
+     */
+    private class FanoutReplicationProxy extends ReplicationProxy {
+
+        private FanoutReplicationProxy() {
+
+        }
+
+        @Override
+        ReplicationMode determineReplicationMode(ReplicationModeAwareShardRouting shardRouting, ShardRouting primaryRouting) {
+            return shardRouting.getShardRouting().isSameAllocation(primaryRouting) == false
+                ? ReplicationMode.LOGICAL_REPLICATION
+                : ReplicationMode.NONE;
+        }
+    }
+
+    /**
+     * This implementation of {@link ReplicationProxy} fans out the replication request to current shard routing basis
+     * the shard routing's replication mode and replication override policy.
+     *
+     * @opensearch.internal
+     */
+    private class ReplicationModeAwareOverrideProxy extends ReplicationProxy {
+
+        private final ReplicationOverridePolicy overridePolicy;
+
+        private ReplicationModeAwareOverrideProxy(ReplicationOverridePolicy overridePolicy) {
+            assert Objects.nonNull(overridePolicy);
+            this.overridePolicy = overridePolicy;
+        }
+
+        @Override
+        ReplicationMode determineReplicationMode(ReplicationModeAwareShardRouting shardRouting, ShardRouting primaryRouting) {
+            ShardRouting currentRouting = shardRouting.getShardRouting();
+
+            // If the current routing is the primary, then it does not need to be replicated
+            if (currentRouting.isSameAllocation(primaryRouting)) {
+                return ReplicationMode.NONE;
+            }
+
+            // If the current routing's replication mode is not NONE, then we return the original replication mode.
+            if (shardRouting.getReplicationMode() != ReplicationMode.NONE) {
+                return shardRouting.getReplicationMode();
             }
+
+            // If the current routing's replication mode is none, then we check for override and return overridden mode.
+            if (Objects.nonNull(overridePolicy)) {
+                return overridePolicy.overriddenMode;
+            }
+
+            // At the end, return NONE.
+            return ReplicationMode.NONE;
+        }
+    }
+
+    /**
+     * Defines the replication override policy which individual {@link TransportReplicationAction} can implement.
+     *
+     * @opensearch.internal
+     */
+    public static class ReplicationOverridePolicy {
```

Comment:
This is for extensibility. Currently we have only the replication mode as a field, but this can expand for future usecases. For the sake of this PR, consider ReplicationOverridePolicy synonymous to ReplicationMode.

Reply:
should we use `ReplicationMode` for now ? Later on we can add abstraction if needed .

Reply:
I think we should keep it. I see cases where depending on flavours of replication mode, we might want to add may be couple of more things - dynamic or static in nature. Since this has huge penetration in tests and other places, i feel we can keep this unless there is a strong reason not to do so.

Reply:
Alright, will make the change.